### ISSUE TITLE
Theming 'select' elements

### ISSUE THREAD

**Issue by LaureeGrd** [shish/shimmie2#827]
#### The problem
This is an example of what's mentioned in #826.
TL;DR: Extensions are using different methods to generate similar code (selectors in this case), but all of them need to be extended by themes that want to make just a simple HTML layout change like adding a wrapper div.

---
#### The possible workarounds
I had to extend two extension themes (and still have to extend many more across all of the site's selectors), copying the code from the original and changing it slightly to include my wrapper element. Another option in this particular case would have been to use JS to add the wrapper maybe, or work with what's there (no wrapper) and try to have consistent CSS (which I already tried as my first 2 approaches at theming, and it was horrible, especially once I added responsiveness to the equation).
If there's any other workaround I missed it, but those will probably have very strong cons to them as well.

---
#### My workaround
As mentioned before, I followed the "shimmie approach" and extended the themes to override the methods that generated those selectors. Here is the comparison:

https://github.com/shish/shimmie2/blob/c4ae68fb471382db6350b82236cf0cc52a2cdf81/ext/pools/theme.php#L383-L390

My extended code including the wrapper div and an icon:
```php
public function get_bulk_pool_selector(array $pools): string
{
    $output = "
    <div class='select-container'>
        <i class='fas fa-ellipsis-h'></i>    
        <select name='bulk_pool_select' required='required'>
        <option></option>
    ";
    foreach ($pools as $pool) {
        $output .= "<option value='" . $pool->id . "' >" . $pool->title . "</option>";
    }
    return $output . "</select></div>";
}
```

And the second one:
https://github.com/shish/shimmie2/blob/c4ae68fb471382db6350b82236cf0cc52a2cdf81/ext/rating/theme.php#L46-L62

My code including the same changes as the other one:
```php
public function get_selection_rater_html(array $selected_options, bool $multiple = false, array $available_options = null): string
{
    $output = "
    <div class='select-container'>
        <i class='fas fa-ellipsis-h'></i>
        <select name='rating".($multiple ? "[]' multiple='multiple'" : "' ")." >
    ";

    $options = Ratings::get_sorted_ratings();

    foreach ($options as $option) {
        if ($available_options!=null && !in_array($option->code, $available_options)) {
            continue;
        }

        $output .= "<option value='".$option->code."' ".
            (in_array($option->code, $selected_options) ? "selected='selected'": "")
            .">".$option->name."</option>";
    }

    return $output."</select></div>";
}
```

---
#### The proposed solution
As mentioned by shish in #793, we could have a function 'build_selector' in core/basepage.php (or where it fits better) that generates and returns these selectors based on parameters like the name, css classes, list of options, etc. And extensions will still take care of their own specific content. Then themes would only have to override that function to change all selectors' layouts.
Themes that override the existing extension themes (or want to do so to modify the code itself) would still be able to do it.

I haven't tried this yet but it seems reasonable. I can code this one but I want to know what others think. If anyone wants to work on it, please discuss how and why you do it that way so I can take the same approach with other similar changes.

**Comment by LaureeGrd** [comment]
And this is how I've started to implement this on my theme now. Notice how changing the extensions to what I did here would mean they work for themes that override build_selector and for themes that use the old approach. That is, I could remove those two files from my theme (and all themes) if we implement this.

## page.class.php
```php
public function build_selector(string $name, array $options, string $attributes, bool $empty_option=false, ?array $selected_options=null): string
{
    $output = "
    <div class='select-container'>
        <i class='fas fa-ellipsis-h'></i>
    ";

    $output .= "<select name='" . $name . "' " . $attributes . ">";

    if ($empty_option) {
        $output .= "<option></option>";
    }

    foreach ($options as $value => $op) {
        if (isset($selected_options) &&  in_array($value, $selected_options)) {
            $output .= "<option value='" . $value . "' selected>" . $op . "</option>";
        } else {
            $output .= "<option value='" . $value . "' >" . $op . "</option>";
        }
    }

    return $output . "</select></div>";
}
```

---
## pools.theme.php
```php
public function get_bulk_pool_selector(array $pools): string
{
    global $page;

    $options = [];

    // This loop wouldn't be necessary if Pools->onBulkActionBlockBuilding queried only the id and title
    // instead of all columns, since it only needs those two values to call this function anyway.
    // I would use array_combine() in the build_selector argument.
    foreach ($pools as $pool) {
        $options[$pool->id] = $pool->title;
    }

    return $page->build_selector("bulk_pool_select", $options, "required", true);        
}
```

---
## rating.theme.php
```php
public function get_selection_rater_html(array $selected_options, bool $multiple = false, array $available_options = null): string
{
    global $page;

    $ratings = Ratings::get_sorted_ratings();
    $options = [];

    foreach ($ratings as $rating) {
        if ($available_options!=null && !in_array($rating->code, $available_options)) {
            continue;
        }

        $options[$rating->code] = $rating->name;
    }

    return $page->build_selector("rating", $options, ($multiple ? "multiple" : ""), false, $selected_options);
}
```

---
A few things to note here:

- My comment on the pools selector function. In fact that function seems unnecessary now, since so little work could be done directly from the call in pools.main.php. Not that I'd do that though, since afaik main.php from extensions is not something you can override with themes, right?
- The $options array could be previously created by a second function that takes 3 arrays (values, option name, selected options) as parameters.
- The $attributes parameter could be incorrectly used by someone extending the selector functions and cause problems. But so does any other variable that injects into HTML like that, and I'm trying to reduce the necessity of overriding so many extensions anyway.

And finally: if the amount of parameters in build_selector keeps growing as a result of adapting more extensions to it, we may have to think of something else. But I like the initial approach, if I say so myself.

**Comment by shish** [comment]
This seems like a good idea :)


> afaik main.php from extensions is not something you can override with themes, right?

Correct - only stuff in `theme.php` can be customised (ideally all HTML would be in themes, and `main.php` would only deal with pure data... but sometimes people (including me) get lazy and put HTML into `main.php` >.>)


> any other variable that injects into HTML like that

In any new code I write, I'm generally using an html-generation library (microhtml) rather than using raw strings - it's a little bit more work, but it means several whole categories of bugs (most importantly security-bugs due to unsafe HTML injection) aren't possible

**Comment by shish** [comment]
Also to expand on the html-generation-library thing - I've started creating some wrappers in `core/util.php` to make sure that common concepts like "a submit button" are done the same way every time (eg `SHM_SUBMIT("blah")` becomes `<input type="submit" value="blah" />`)

I wonder if these functions should actually live in the base theme, so that custom themes can override them 🤔

**Comment by LaureeGrd** [comment]
> Correct - only stuff in `theme.php` can be customised (ideally all HTML would be in themes, and `main.php` would only deal with pure data...

Right, so I will work on moving HTML to `theme.php` where necessary first, then we can more comfortably apply the new ideas. First offender I found:
https://github.com/shish/shimmie2/blob/c4ae68fb471382db6350b82236cf0cc52a2cdf81/ext/setup/main.php#L236-L257

> In any new code I write, I'm generally using an html-generation library (microhtml) rather than using raw strings

Yes I noticed those in the ImageInfo table already! IIRC it was another offender in terms of generating its HTML in `main.php`. Tbh I didn't have the time to get my head around microhtml yet but it would be a good idea to use that when possible then.

**Comment by LaureeGrd** [comment]
>Also to expand on the html-generation-library thing - I've started creating some wrappers in `core/util.php` to make sure that common concepts like "a submit button" are done the same way every time

That's a good one too. I wonder which elements should be done in that way and what's best to keep simple. In my experience, any helper function that returns HTML tags that are not closed is problematic. Example:

https://github.com/shish/shimmie2/blob/c4ae68fb471382db6350b82236cf0cc52a2cdf81/core/util.php#L703

This one generates a `<form>`, which is useful, but never closes it. Having such helpers for very specific things is ok but I can imagine too many of those "partial" HTML helpers would be confusing to work with.

> I wonder if these functions should actually live in the base theme, so that custom themes can override them 🤔

I think so. Being able to override simple functions without having to rewrite a lot of code is always better for me, mainly because it makes themes more resistant to changes in the base code. But also because it makes creating themes faster in general. I found myself spending too much time adapting things that I wouldn't have touched otherwise rather than spend it designing my layout. (I guess I'm still doing that with this lol)

I do have a question, though: Would that approach (with the small helpers) replace the one I posted here? That is, would we still have a function that combines said helpers to create entire blocks in `basepage.php` or `basethemelet.php` like the `build_selector` example, or would extensions directly use helpers on their `theme.php` functions?

If we kept both, themes could use a mix of 3 options according to how specific the layout change needs to be:

1. Override the new generalized layout functions that affect all extensions.
2. Override those helpers to change specific HTML tags across all the site.
3. Override an extension's specific layout, like until now.

I like this approach more than having just the new functions or just the helpers.
